tests/test_download.py

```diff
@@ -168,81 +172,188 @@ def test_download_url_to_tempfileobj_and_urls(self):
-      download.unsafe_download(https_url, target_data_length)
+      # Trust the certfile that happens to use a different hostname than we
+      # will expect.
+      os.environ['REQUESTS_CA_BUNDLE'] = bad_cert_fname
+
+      # Try connecting to the server process with the bad cert while trusting
+      # the bad cert. Expect failure because even though we trust it, the
+      # hostname we're connecting to does not match the hostname in the cert.
+      logger.info('Trying https download of target file: ' + bad_https_url)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.safe_download(bad_https_url, target_data_length)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.unsafe_download(bad_https_url, target_data_length)
+
+      # Try connecting to the server processes with the good certs while not
+      # trusting the good certs (trusting the bad cert instead). Expect failure
+      # because even though the server's cert file is otherwise OK, we don't
+      # trust it.
+      print('Trying https download of target file: ' + good_https_url)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.safe_download(good_https_url, target_data_length)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.unsafe_download(good_https_url, target_data_length)
+
+      print('Trying https download of target file: ' + good2_https_url)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.safe_download(good2_https_url, target_data_length)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.unsafe_download(good2_https_url, target_data_length)
+
+      # Configure environment to now trust the certfile that is expired.
+      os.environ['REQUESTS_CA_BUNDLE'] = expired_cert_fname
+
+      # Try connecting to the server process with the expired cert while
+      # trusting the expired cert. Expect failure because even though we trust
+      # it, it is expired.
+      logger.info('Trying https download of target file: ' + expired_https_url)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.safe_download(expired_https_url, target_data_length)
+      with self.assertRaises(requests.exceptions.SSLError):
+        download.unsafe_download(expired_https_url, target_data_length)
+
+      # Try connecting to the server processes with the good certs while
+      # trusting the appropriate good certs. Expect success.
+      # Note: running these OK downloads at the top of this try section causes
+      #       a failure in a previous assertion: retrieving the same good URL
+      #       again after no longer "trusting" the good certfile still succeeds
+      #       if we had previously succeeded in retrieving that same URL while
+      #       still trusting the good cert. Perhaps it's a caching issue....?
+      #       I'm not especially concerned yet, but take note for later....
```

Comment:
It's probably worth checking if `requests` does what you expect it to (i.e. reload the CA bundle from the environment variable). A [quick glance suggests](https://github.com/requests/requests/blob/ff0c325014f817095de35013d385e137b111d6e8/requests/sessions.py#L698) that a new `Session` must be created for a reload to happen. (Are you sure that the explanations for the expected failures above are well-grounded? Are the bad etc. certs ever seen by `requests`?)

Reply:
Thanks! I think I saw a bit of behavior supporting this notion, but there was enough varied testing that I don't think the tests weren't succeeding at what they were doing. I'm curious, but this was easily solved by purging the sessions when the certificate expectations changed, so I've removed the possibility for this issue and removed a related comment now that it's gone.